// updated-node-msmq/proxy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.queueProxy = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _edgeJs = require('edge-js');

var _edgeJs2 = _interopRequireDefault(_edgeJs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var baseOptions = {
	assemblyFile: _path2.default.join(__dirname, 'MSMQLib.dll'),
	typeName: 'MSMQLib.MSMQInterface'
};

function getMethod(methodName) {
	return _edgeJs2.default.func((0, _assign2.default)({}, baseOptions, { methodName: methodName }));
}

var queueProxy = exports.queueProxy = {
	exists: getMethod('ExistsQueue'),
	create: getMethod('CreateQueue'),
	send: getMethod('SendMessage'),
	receive: getMethod('ReceiveMessages'),
	list: getMethod('GetAllMessages'),
	clear: getMethod('PurgeQueue'),
	connectRemote: getMethod('ConnectRemote')
};
