// updated-node-msmq/queue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _events = require('events');

var _proxy = require('./proxy');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Queue = function (_EventEmitter) {
	(0, _inherits3.default)(Queue, _EventEmitter);

	function Queue(path) {
		(0, _classCallCheck3.default)(this, Queue);

		var _this = (0, _possibleConstructorReturn3.default)(this, (Queue.__proto__ || (0, _getPrototypeOf2.default)(Queue)).call(this));

		_this.path = path;
		return _this;
	}

	(0, _createClass3.default)(Queue, [{
		key: 'startReceiving',
		value: function startReceiving() {
			var _this2 = this;

			if (this.receiving) {
				throw new Error('Already receiving messages from this queue');
			}

			this.receiving = true;

			return _proxy.queueProxy.receive({
				path: this.path,
				receive: function receive(message) {
					_this2.emit('receive', message);
				}
			});
		}
	}, {
		key: 'send',
		value: function send(message, cb) {
			var formattedMessage = (0, _stringify2.default)(message);

			return _proxy.queueProxy.send({
				path: this.path,
				message: formattedMessage
			}, cb);
		}
	}, {
		key: 'getAllMessages',
		value: function getAllMessages() {
			return _proxy.queueProxy.list(this.path, true);
		}
	}, {
		key: 'purge',
		value: function purge() {
			_proxy.queueProxy.clear(this.path, true);
		}
	}], [{
		key: 'existsQueue',
		value: function existsQueue(path) {
			return _proxy.queueProxy.exists(path, true);
		}
	}, {
		key: 'createQueue',
		value: function createQueue(path) {
			var result = _proxy.queueProxy.create(path, true);

			if (!result) {
				throw new Error('Queue already exists');
			}

			return new Queue(path);
		}
	}, {
		key: 'openOrCreateQueue',
		value: function openOrCreateQueue(path) {
			_proxy.queueProxy.create(path, true);
			return new Queue(path);
		}
	}, {
		key: 'connectToRemoteQueue',
		value: function connectToRemoteQueue(path) {
			_proxy.queueProxy.connectRemote(path, true);
			return new Queue(path);
		}
	}]);
	return Queue;
}(_events.EventEmitter);

exports.default = Queue;
module.exports = exports['default'];
